clash-cores/src/Clash/Cores/Xilinx/Xpm/Cdc/Gray.hs

```diff
@@ -0,0 +1,89 @@
+{-|
+  Copyright   :  (C) 2023, Google LLC
+  License     :  BSD2 (see the file LICENSE)
+  Maintainer  :  QBayLogic B.V. <devops@qbaylogic.com>
+-}
+{-# LANGUAGE NamedFieldPuns #-}
+
+{-# OPTIONS_GHC -fno-do-lambda-eta-expansion #-}
+
+module Clash.Cores.Xilinx.Xpm.Cdc.Gray
+  ( xpmCdcGray
+  , XpmCdcGrayConfig(..)
+  , xpmCdcGrayWith
+  ) where
+
+import GHC.Stack (HasCallStack)
+
+import Clash.Explicit.Prelude
+
+import Clash.Cores.Xilinx.Xpm.Cdc.Gray.Internal (xpmCdcGray#)
+
+-- | Synchronizes an 'Unsigned' from the source clock domain to the destination
+-- clock domain using Gray code. It instantiates Xilinx's @XPM_CDC_GRAY@, so no
+-- additional constraint definitions are needed. In order to synchronize data
+-- from the source domain to the destination domain, consecutive inputs need to
+-- be stable, successors (+1), or predecessors (-1). Overflows are okay. That is,
+-- 'maxBound' can be followed by zero, and vice versa. If these invariants are
+-- violated, the component might not transfer data correctly until the source clock
+-- has sampled the input once, followed by /4/ samples in the destination
+-- domain.
```

Comment:
If there are 4 synchronizer stages, wouldn't this be _5_ samples? The first sample of the first synchronizer stage after an invariant violation might be corrupted, but, when the invariant is upheld once again, the second sample will be clean. Followed by three more synchronizer stages before it is observable.

Also, let's make sure the user understands that data will be corrupted until they uphold the invariants again, because a naive reading of this documentation might lead to the conclusion that as long as you wait _n_ samples for certain _n_, data will be clean.

```suggestion
-- 'maxBound' can be followed by zero, and vice versa. If these invariants are
-- violated, the component might not transfer data correctly until the invariant
-- is upheld again and the source clock has sampled the input once, followed by
-- /4/ samples in the destination domain.
```

Note that the extra sample is now part of the text :-).